### PR TITLE
Fix closing the dialogs

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Dialog must inform somehow the caller when the dialog is closed bu clicking 'Cancel' button.

Issue https://github.com/theia-ide/theia/issues/2400
The behavior is needed for https://github.com/theia-ide/theia/issues/2153

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

### CODE REVIEW

**Comment by vitaliy-guliy** [comment]
Great! I will try to rework it and call `reject()` when closing. Thanks!

**Comment by akosyakov** [comment]
I've tried to open a new file dialog and close it, and got following errors:
screen shot 2018-07-23 at 09 45 08

We should make sure that all existing clients handle it properly, otherwise, the console will be full of errors. Also, we should reject with a special error as in `saveable` to distinguish normal closing from other cases. We could try to make use of (onunhandledrejection)[https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onunhandledrejection] and log with debug level if an error is canceled generally and otherwise with an error.

**Comment by vitaliy-guliy** [comment]
Sorry, I will check again. It was working fine on my side.

**Comment by akosyakov** [comment]
Have you looked at installing global https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onunhandledrejection to allow clients skipping handling of closing? The issue is that it is tedious to add catch case always and clients had to do it but not enforced.

**Comment by vitaliy-guliy** [comment]
Yes, I agree with you that it is tedious to add catch cases always to handle closing the dialogs.
I added handling of unhandled rejections to `FrontendApplication`

unhandled rejection

It works fine. But I'm not sure we need to catch and hide all unhandled rejections without notifying / writing to the console.
Will I proceed with this handler?

**Comment by akosyakov** [comment]
> It works fine. But I'm not sure we need to catch and hide all unhandled rejections without notifying / writing to the console. 

We should not reject with `undefined` but with a special `canceled` error as in `saveable.ts`. In the handler, If it is a canceled error we should log it with `console.debug` and suppress event propagation. For other errors and `undefined` it should work as before.


> Will I proceed with this handler?

I like the handler. cc @svenefftinge what do you think?

**Comment by vitaliy-guliy** [comment]
I would say calling the `resolve` and passing `undefined` value is more convenient for the developer.
Especially when you use the dialog in `asyc` method and wait for some value.
When you got `undeifned` you understand the dialog has been closed and nothing was accepted.
Yes, then dialog will not reject the promise when closing but do we really need to consider the pattern `resolve/reject` ??

**Comment by vitaliy-guliy** [comment]
Let you see an examples

```javascript
async shouldOverwrite(file: FileStat, stat: FileStat): Promise<boolean> {
    const dialog = new ConfirmDialog({
        title: `The file '${file.uri}' has been changed on the file system.`,
        msg: 'Do you want to overwrite the changes made on the file system?',
        ok: 'Yes',
        cancel: 'No'
    });
    return await dialog.open();
}
```

The code above looks better for the developer then following

```javascript
async shouldOverwrite(file: FileStat, stat: FileStat): Promise<boolean> {
    try {
        const dialog = new ConfirmDialog({
            title: `The file '${file.uri}' has been changed on the file system.`,
            msg: 'Do you want to overwrite the changes made on the file system?',
            ok: 'Yes',
            cancel: 'No'
        });
        return await dialog.open();
    } catch (error) {
        if (error) {
            console.error(error);
        }
    }
    return false;
}
```

**Comment by akosyakov** [comment]
> Let you see an examples

`boolean` as a return type with `false` on close is a bad example, consider if on close `true` is expected or an empty array, then with true on close:
```ts
async shouldOverwrite(file: FileStat, stat: FileStat): Promise<boolean> {
    const dialog = new ConfirmDialog({
        title: `The file '${file.uri}' has been changed on the file system.`,
        msg: 'Do you want to overwrite the changes made on the file system?',
        ok: 'Yes',
        cancel: 'No'
    });
    const result = await dialog.open();
    if (typeof result === 'boolean') {
        return result;
    }
    return true;
}
```

It is not better than `catch(e) {}`.

the bottom line: with rejecting with a special error and global handling of it, we can leave all code as it is.

**Comment by akosyakov** [comment]
Also, it is semantically bogus, for example, if we rewrite `saveable` with resolving to `undefined` then this code will be bogus: https://github.com/theia-ide/theia/blob/c88e7d9b0eea00fe36e6b5c662224e7dfaa659c7/packages/core/src/browser/saveable.ts#L78-L89

It will close a widget without saving instead of aborting the whole operation.

**Comment by akosyakov** [comment]
Although, thinking about it more, `saveable` could be rewritten like:
```ts
const result = await shouldSave(saveable, widget);
if (typeof result === 'boolean') {
    if (result) {
        await Saveable.save(widget);
    }
    close();
}
```

@vitaliy-guliy ok, let's give resolving to `undefined` a try, but please:
- make sure that types are correct, i.e. `open` should return `T | undefined` where `T` is allowed values
- review existing clients that they changed properly, like `saveable` should be rewritten as above

**Comment by vitaliy-guliy** [comment]
Considering `await shouldSave(...)` always returns `Promise<boolean>` and `ShouldSaveDialog.open()` returns the same we can refactor the case and remove type checking
```javascript
try {
    if (await shouldSave(saveable, widget)) {
        await Saveable.save(widget);
    }
    close();
} finally {
    closing = false;
}
```
Anyway I will check it again.

**Comment by akosyakov** [comment]
> Considering await shouldSave(...) always returns Promise<boolean> and ShouldSaveDialog.open() returns the same we can refactor the case and remove type checking

please reread https://github.com/theia-ide/theia/pull/2401#issuecomment-407360104, `false` and `undefined` should be handled differently:
- `false` - a widget should be closed without saving
- `undefined` - a widget should not be closed at all

**Comment by vitaliy-guliy** [comment]
Yes, indeed. I missed it.

**Comment by vitaliy-guliy** [comment]
@akosyakov please give a look at the improvements. The logic in `saveable` and using `ShouldSaveDialog` has been properly checked.

File: packages/filesystem/src/browser/file-tree/file-tree-model.ts [hunk 1]
```diff
@@ -145,7 +145,7 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
             ok: 'Yes',
             cancel: 'No'
         });
-        return dialog.open();
+        return await dialog.open() ? true : false;
```
Comment (akosyakov) [inline packages/filesystem/src/browser/file-tree/file-tree-model.ts]: `!!await dialog.open()` wonder whether such will work?

**Pull request merged**